### PR TITLE
implement proper ECMP route handling

### PR DESCRIPTION
**Pull request by KanjiMonster**
Implement ECMP route handling:

The current ECMP route handling implementation has seveal drawbacks:

* it assumes that all nexthops are reachable
* it does not handle nexthops becoming (un)reachable
* it uses a `set<l3_interface_id>` to ecmp group id, so if nexthops change their availability during route updates, ecmp groups may become orphaned
* it does not recycle ecmp group ids, so eventually it runs out

Fix this by implementing a `set<nh_stub>` to ecmp group id matching, for stable groups regardless of reachability of nexthops, and make `nl_l3::nh_{un}reachable_notification()` update ecmp routes, and not just mark them as (un)routable.

Finally, change the ECMP group id creation/destruction and id allocation being done in `of-dpa::controller`, like we already do for l3 interface ids, and recycle freed one. This ensures we do not eventually run out of ids.

### CODE REVIEW

**Comment by KanjiMonster** [comment]
Depends on https://github.com/bisdn/basebox/pull/443

File: src/of-dpa/controller.cc [hunk 1]
```diff
@@ -1496,10 +1536,10 @@ int controller::l3_ecmp_remove(uint32_t l3_ecmp_id) noexcept {
   try {
     rofl::crofdpt &dpt = set_dpt(dptid, true);
 
-    l3_ecmp_id = fm_driver.group_id_l3_ecmp(l3_ecmp_id);
     dpt.send_group_mod_message(
         rofl::cauxid(0),
-        fm_driver.disable_group_l3_ecmp(dpt.get_version(), l3_ecmp_id));
+        fm_driver.disable_group_l3_ecmp(
+            dpt.get_version(), fm_driver.group_id_l3_ecmp(_ecmp_interface_id)));
```
Comment (rubensfig) [inline src/of-dpa/controller.cc:1542]: Compilation fails with

```
Jun 13 09:25:16 | ../git/src/of-dpa/controller.cc:1542:59: error: '_ecmp_interface_id' was not declared in this scope; did you mean 'ecmp_interface_id'?
Jun 13 09:25:16 |  1542 |             dpt.get_version(), fm_driver.group_id_l3_ecmp(_ecmp_interface_id)));
Jun 13 09:25:16 |       |                                                           ^~~~~~~~~~~~~~~~~~
Jun 13 09:25:16 |       |                                                           ecmp_interface_id
```

Comment (KanjiMonster) [reply src/of-dpa/controller.cc]: weird it didn't happen for me, I must have flubbed the testing somehow.

Should be fixed.

**Comment by rubensfig** [comment]
After running switch-testing tests, the following state was encountered.

The installed flow entries has one extra "default route"

```
Table ID 30 (Unicast Routing):   Retrieving all entries. Max entries = 73728, Current entries = 5.
**--  etherType = 0x86dd vrf:mask = 0x0000:0x0000 dstIp4 = 0.0.0.0/0.0.0.0 dstIp6 = ::/:: | GoTo = 60 (ACL Policy) groupId = 0x20000001 | priority = 2 hard_time = 0 idle_time = 0 cookie = 23**
--  etherType = 0x86dd vrf:mask = 0x0000:0x0000 dstIp4 = 0.0.0.0/0.0.0.0 dstIp6 = fe80::/ffff:ffff:ffff:ffff:: | GoTo = 60 (ACL Policy) outPort = CONTROLLER (Reserved)  | priority = 2 hard_time = 0 idle_time = 0 cookie = 24
--  etherType = 0x86dd vrf:mask = 0x0000:0x0000 dstIp4 = 0.0.0.0/0.0.0.0 dstIp6 = 2001:db8:0:100::/ffff:ffff:ffff:ffff:: | GoTo = 60 (ACL Policy) groupId = 0x20000003 | priority = 2 hard_time = 0 idle_time = 0 cookie = 29
--  etherType = 0x86dd vrf:mask = 0x0000:0x0000 dstIp4 = 0.0.0.0/0.0.0.0 dstIp6 = 2001:db8:0:101::/ffff:ffff:ffff:ffff:: | GoTo = 60 (ACL Policy) groupId = 0x20000004 | priority = 2 hard_time = 0 idle_time = 0 cookie = 30
--  etherType = 0x86dd vrf:mask = 0x00fe:0xffff dstIp4 = 0.0.0.0/0.0.0.0 dstIp6 = ::/:: | GoTo = 60 (ACL Policy) groupId = 0x70000003 | priority = 2 hard_time = 0 idle_time = 0 cookie = 25

```

**Comment by rubensfig** [comment]
Moreover, while testing, the switch state was reset, by running `basebox-change-config -l baseboxd`.

The following state is now observed.

```
Table ID 30 (Unicast Routing):   Retrieving all entries. Max entries = 73728, Current entries = 7.
--  etherType = 0x86dd vrf:mask = 0x0000:0x0000 dstIp4 = 0.0.0.0/0.0.0.0 dstIp6 = ::/:: | GoTo = 60 (ACL Policy) groupId = 0x20000001 | priority = 2 hard_time = 0 idle_time = 0 cookie = 24
--  etherType = 0x86dd vrf:mask = 0x0000:0x0000 dstIp4 = 0.0.0.0/0.0.0.0 dstIp6 = fe80::/ffff:ffff:ffff:ffff:: | GoTo = 60 (ACL Policy) outPort = CONTROLLER (Reserved)  | priority = 2 hard_time = 0 idle_time = 0 cookie = 14
--  etherType = 0x86dd vrf:mask = 0x0000:0x0000 dstIp4 = 0.0.0.0/0.0.0.0 dstIp6 = 2001:db8:0:100::/ffff:ffff:ffff:ffff:: | GoTo = 60 (ACL Policy) groupId = 0x20000004 | priority = 2 hard_time = 0 idle_time = 0 cookie = 32
--  etherType = 0x86dd vrf:mask = 0x0000:0x0000 dstIp4 = 0.0.0.0/0.0.0.0 dstIp6 = 2001:db8:0:101::/ffff:ffff:ffff:ffff:: | GoTo = 60 (ACL Policy) groupId = 0x20000003 | priority = 2 hard_time = 0 idle_time = 0 cookie = 31
--  etherType = 0x86dd vrf:mask = 0x00fe:0xffff dstIp4 = 0.0.0.0/0.0.0.0 dstIp6 = ::/:: | GoTo = 60 (ACL Policy) groupId = 0x70000003 | priority = 2 hard_time = 0 idle_time = 0 cookie = 28
--  etherType = 0x86dd vrf:mask = 0x00fe:0xffff dstIp4 = 0.0.0.0/0.0.0.0 dstIp6 = 2001:db8:0:100::/ffff:ffff:ffff:ffff:: | GoTo = 60 (ACL Policy) groupId = 0x70000001 | priority = 2 hard_time = 0 idle_time = 0 cookie = 29
--  etherType = 0x86dd vrf:mask = 0x00fe:0xffff dstIp4 = 0.0.0.0/0.0.0.0 dstIp6 = 2001:db8:0:101::/ffff:ffff:ffff:ffff:: | GoTo = 60 (ACL Policy) groupId = 0x70000001 | priority = 2 hard_time = 0 idle_time = 0 cookie = 30
```

```
groupId = 0x70000001 (L3 ECMP, Index = 1): duration: 47, refCount:2
        bucketIndex = 0: referenceGroupId = 0x20000001 
        bucketIndex = 1: referenceGroupId = 0x20000002 
groupId = 0x70000003 (L3 ECMP, Index = 3): duration: 45, refCount:1
        bucketIndex = 0: referenceGroupId = 0x20000001 
        bucketIndex = 1: referenceGroupId = 0x20000002 
        bucketIndex = 2: referenceGroupId = 0x20000003 
        bucketIndex = 3: referenceGroupId = 0x20000004 
```

With  the following state present in the Linux kernel.
````
2001:db8:0:100::/64 nhid 86 via fe80::ec4:7aff:fe9c:29fc dev port5 proto bgp metric 20 pref medium
2001:db8:0:101::/64 nhid 77 via fe80::ec4:7aff:fe9c:27d3 dev port4 proto bgp metric 20 pref medium
fe80::/64 dev enp0 proto kernel metric 256 pref medium
fe80::/64 dev port4 proto kernel metric 256 pref medium
fe80::/64 dev port5 proto kernel metric 256 pref medium
fe80::/64 dev port53 proto kernel metric 256 pref medium
fe80::/64 dev port54 proto kernel metric 256 pref medium
default proto ra metric 1024 expires 27sec pref medium
        nexthop via fe80::e201:a6ff:fea9:9a55 dev port54 weight 1 
        nexthop via fe80::e201:a6ff:fea9:9a54 dev port53 weight 1 
        nexthop via fe80::ec4:7aff:fe9c:27d3 dev port4 weight 1 
        nexthop via fe80::ec4:7aff:fe9c:29fc dev port5 weight 1 
```

**Comment by KanjiMonster** [comment]
which platform did you use? Also you don't need to use `basebox-change-config` anymore, just doing a `systemctl restart baseboxd` is enough, as `baseboxd` takes care of resetting the OpenFlow configuration on startup nowadays.

**Comment by rubensfig** [comment]
Tested on the 4630 setup. Thanks for the tip!

**Comment by jklare** [comment]
> which platform did you use? Also you don't need to use `basebox-change-config` anymore, just doing a `systemctl restart baseboxd` is enough, as `baseboxd` takes care of resetting the OpenFlow configuration on startup nowadays.

can we drop the "basebox-change-config" script ?

**Comment by rubensfig** [comment]
I thought the `basebox-change-config` script was a convenience tool to use the `ryu` controller. If there are no users using this controller, then we could do it.

Otherwise changing that script to just restart baseboxd should be fine.

**Comment by KanjiMonster** [comment]
`basebox-change-config` is still needed for (runtime) switching between `baseboxd` and `ryu`, since `ryu` does not (and cannot for some parts) reset the configuration of OF-DPA.

We could optimize it for the `-l (baseboxd)` case to just restart baseboxd in this case, but all other cases will need a full restart of all components.

**Comment by KanjiMonster** [comment]
> Moreover, while testing, the switch state was reset, by running `basebox-change-config -l baseboxd`.
> 
> The following state is now observed.
> 
> ```
> Table ID 30 (Unicast Routing):   Retrieving all entries. Max entries = 73728, Current entries = 7.
> --  etherType = 0x86dd vrf:mask = 0x0000:0x0000 dstIp4 = 0.0.0.0/0.0.0.0 dstIp6 = ::/:: | GoTo = 60 (ACL Policy) groupId = 0x20000001 | priority = 2 hard_time = 0 idle_time = 0 cookie = 24
> --  etherType = 0x86dd vrf:mask = 0x0000:0x0000 dstIp4 = 0.0.0.0/0.0.0.0 dstIp6 = fe80::/ffff:ffff:ffff:ffff:: | GoTo = 60 (ACL Policy) outPort = CONTROLLER (Reserved)  | priority = 2 hard_time = 0 idle_time = 0 cookie = 14
> --  etherType = 0x86dd vrf:mask = 0x0000:0x0000 dstIp4 = 0.0.0.0/0.0.0.0 dstIp6 = 2001:db8:0:100::/ffff:ffff:ffff:ffff:: | GoTo = 60 (ACL Policy) groupId = 0x20000004 | priority = 2 hard_time = 0 idle_time = 0 cookie = 32
> --  etherType = 0x86dd vrf:mask = 0x0000:0x0000 dstIp4 = 0.0.0.0/0.0.0.0 dstIp6 = 2001:db8:0:101::/ffff:ffff:ffff:ffff:: | GoTo = 60 (ACL Policy) groupId = 0x20000003 | priority = 2 hard_time = 0 idle_time = 0 cookie = 31
> --  etherType = 0x86dd vrf:mask = 0x00fe:0xffff dstIp4 = 0.0.0.0/0.0.0.0 dstIp6 = ::/:: | GoTo = 60 (ACL Policy) groupId = 0x70000003 | priority = 2 hard_time = 0 idle_time = 0 cookie = 28
> --  etherType = 0x86dd vrf:mask = 0x00fe:0xffff dstIp4 = 0.0.0.0/0.0.0.0 dstIp6 = 2001:db8:0:100::/ffff:ffff:ffff:ffff:: | GoTo = 60 (ACL Policy) groupId = 0x70000001 | priority = 2 hard_time = 0 idle_time = 0 cookie = 29
> --  etherType = 0x86dd vrf:mask = 0x00fe:0xffff dstIp4 = 0.0.0.0/0.0.0.0 dstIp6 = 2001:db8:0:101::/ffff:ffff:ffff:ffff:: | GoTo = 60 (ACL Policy) groupId = 0x70000001 | priority = 2 hard_time = 0 idle_time = 0 cookie = 30
> ```
> 
> ```
> groupId = 0x70000001 (L3 ECMP, Index = 1): duration: 47, refCount:2
>         bucketIndex = 0: referenceGroupId = 0x20000001 
>         bucketIndex = 1: referenceGroupId = 0x20000002 
> groupId = 0x70000003 (L3 ECMP, Index = 3): duration: 45, refCount:1
>         bucketIndex = 0: referenceGroupId = 0x20000001 
>         bucketIndex = 1: referenceGroupId = 0x20000002 
>         bucketIndex = 2: referenceGroupId = 0x20000003 
>         bucketIndex = 3: referenceGroupId = 0x20000004 
> ```
> 
> With the following state present in the Linux kernel.
> 
> ```
> 2001:db8:0:100::/64 nhid 86 via fe80::ec4:7aff:fe9c:29fc dev port5 proto bgp metric 20 pref medium
> 2001:db8:0:101::/64 nhid 77 via fe80::ec4:7aff:fe9c:27d3 dev port4 proto bgp metric 20 pref medium
> fe80::/64 dev enp0 proto kernel metric 256 pref medium
> fe80::/64 dev port4 proto kernel metric 256 pref medium
> fe80::/64 dev port5 proto kernel metric 256 pref medium
> fe80::/64 dev port53 proto kernel metric 256 pref medium
> fe80::/64 dev port54 proto kernel metric 256 pref medium
> default proto ra metric 1024 expires 27sec pref medium
>         nexthop via fe80::e201:a6ff:fea9:9a55 dev port54 weight 1 
>         nexthop via fe80::e201:a6ff:fea9:9a54 dev port53 weight 1 
>         nexthop via fe80::ec4:7aff:fe9c:27d3 dev port4 weight 1 
>         nexthop via fe80::ec4:7aff:fe9c:29fc dev port5 weight 1 
> ```
> ```

With the ECMP changes applied, I cannot reproduce this, and the bgp-ipv6-multipath test succeeds even with the added flow/group table checks.

Without the ECMP changes I see exactly this behaviour though. Please double check that you actually tested these changes.

**Comment by rubensfig** [comment]
Tested again with the correct version installed, all looks good. 

PR ready for undraft, thank you for checking again.

**Review by rubensfig — APPROVED** [review]
LGTM!

**Pull request merged**